pkg/skaffold/build/kaniko/sources/localdir.go

```diff
@@ -0,0 +1,129 @@
+	logrus.Infof("storing buildcontext tarball at %s", g.tarPath)
+	f, err := os.Create(g.tarPath)
+	if err != nil {
+		return "", errors.Wrap(err, "creating temporary buildcontext tarball")
+	}
+	defer f.Close()
+	err = docker.CreateDockerTarGzContext(f, artifact.Workspace, artifact.DockerArtifact)
+	context := fmt.Sprintf("dir://%s", constants.DefaultKanikoEmptyDirMountPath)
+	return context, err
+}
+
+// Pod returns the pod template to ModifyPod
+func (g *LocalDir) Pod(cfg *latest.KanikoBuild, args []string) *v1.Pod {
+	p := podTemplate(cfg, args)
+	// Include the emptyDir volume and volume source in both containers
+	v := v1.Volume{
+		Name: constants.DefaultKanikoEmptyDirName,
+		VolumeSource: v1.VolumeSource{
+			EmptyDir: &v1.EmptyDirVolumeSource{},
+		},
+	}
+	vm := v1.VolumeMount{
+		Name:      constants.DefaultKanikoEmptyDirName,
+		MountPath: constants.DefaultKanikoEmptyDirMountPath,
+	}
+	// Generate the init container, which will run until the /tmp/complete file is created
+	ic := v1.Container{
+		Name:  initContainer,
+		Image: constants.DefaultDebianImage,
+		Args: []string{"sh", "-c", `while true; do
+	sleep 1; if [ -f /tmp/complete ]; then break; fi
+done`},
+		VolumeMounts: []v1.VolumeMount{vm},
+	}
+
+	p.Spec.InitContainers = []v1.Container{ic}
+	p.Spec.Containers[0].VolumeMounts = append(p.Spec.Containers[0].VolumeMounts, vm)
+	p.Spec.Volumes = append(p.Spec.Volumes, v)
+	return p
+}
+
+// ModifyPod first copies over the buildcontext tarball into the init container tmp dir via kubectl cp
+// Via kubectl exec, we extract the tarball to the empty dir
+// Then, via kubectl exec, create the /tmp/complete file via kubectl exec to complete the init container
+func (g *LocalDir) ModifyPod(p *v1.Pod) error {
+	// Wait for the init container to start running
+	if err := waitForInitContainer(p); err != nil {
+		return errors.Wrap(err, "waiting for init container")
+	}
+	// Copy over the buildcontext tarball into the init container
+	copy := exec.Command("kubectl", "cp", g.tarPath, fmt.Sprintf("%s:/%s", p.Name, g.tarPath), "-c", initContainer, "-n", p.Namespace)
+	if err := util.RunCmd(copy); err != nil {
+		return errors.Wrap(err, "copying buildcontext into init container")
+	}
+	// Next, extract the buildcontext to the empty dir
+	extract := exec.Command("kubectl", "exec", p.Name, "-c", initContainer, "-n", p.Namespace, "--", "tar", "-xzf", g.tarPath, "-C", constants.DefaultKanikoEmptyDirMountPath)
+	if err := util.RunCmd(extract); err != nil {
+		return errors.Wrap(err, "extracting buildcontext to empty dir")
+	}
+	// Generate a file to successfully terminate the init container
+	file := exec.Command("kubectl", "exec", p.Name, "-c", initContainer, "-n", p.Namespace, "--", "touch", "/tmp/complete")
+	return util.RunCmd(file)
+}
+
+// waitForInitContainer tries to exec into the init container until it doesn't fail
+func waitForInitContainer(p *v1.Pod) error {
```

Comment:
Query the apiserver for init container status instead?

Reply:
There’s some example of pod watchers in the code. 

We should remove all the PollImmediates when we get a chance. I added those a long time ago in minikube before I did it right.

Reply:
yah that's way better, added it in, thanks!